Remove support for rocBlas from TE

alextmagro opened the pull request:
# Description

rocBlas build path fails for ROCm 7.0+. As all AI development will be in hipBlasLt moving forward, we are removing support for rocBlas

Fixes # (issue)

#12604 [TE] Disable rocblas for ROCm7.0


## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Infra/Build change

## Changes

- Removed usage of USE_ROCBLAS and USE_HIPBLASLT flags
- Removed conditional logic for hipBlasLt/rocBlas and left only hipblas relevant code
- Removed rocBlas specific tests

Commit:
Remove support for rocBlas from TE

ipanfilo commented on tests/pytorch/test_numerics.py:
```diff
@@ -825,13 +821,7 @@ def test_gpt_checkpointing(dtype, bs, model):
-            tols["atol"] = 5e-5
-        else:
-            # should be perfect match with hipblaslt and non-ck attention
+        if len(tols) == 0:
```
The code seems void, let's remove it

ipanfilo commented on transformer_engine/common/gemm/rocm_gemm.cu:
```diff
@@ -1402,505 +1068,40 @@ void hipblaslt_gemm(const Tensor *inputA,
+                  inputCounter, stream,
+                  handle);
+
+  return;
```
Not needed, there is no any code after this return

ipanfilo commented on transformer_engine/common/gemm/rocm_gemm.cu:
```diff
@@ -1402,505 +1068,40 @@ void hipblaslt_gemm(const Tensor *inputA,
-      static hipblasLtHandle_t hipblaslt_handles[num_streams];
-      std::call_once(init_flag, init_hipblaslt_handles, hipblaslt_handles);
+{  
+  // ROCBLAS support has been removed. HIPBLASLT is the only GEMM backend supported.
```
Does it make sense to combine hipblaslt_gemm and cublas_gemm then?

wangye805 replied:
I agree with @ipanfilo 's suggestion. In fact, I was thinking we can remove this rocm_gemm.cu and merge everything inside it to cublaslt_gemm.cu since our flow is greatly simplified. A lot of codes can share with hipified cuda code. Like get_hipblaslt_type(). In v2.1, we will have more codes to share, like gemm_param and canonicalize_gemm_input: https://github.com/ROCm/TransformerEngine/blob/3b36fc5217805b16b819f2af52e27db8ba7088bc/transformer_engine/common/gemm/rocm_gemm.cu#L90

alextmagro replied:
We discussed moving this to a different ticket.

ipanfilo commented on tests/pytorch/test_gemm_autotune.py:
Update copyright date

ipanfilo commented on ci/pytorch.sh:
```diff
@@ -97,41 +97,37 @@ install_prerequisites
 
-for _gemm in hipblaslt rocblas; do
-    configure_gemm_env $_gemm || continue
+_gemm="hipblaslt"
```
Since there is only one GEMM backend, _gemm is not needed and can be removed

ipanfilo replied:
By removing _gemm I did not mean replacing it with hardcoded value but remove it's usage. See how it is made in jax.sh that does not have GEMM usage

alextmagro replied:
I think I made the changes you are referring to in my follow-up commit, can you double check? Thanks.

ipanfilo commented on ci/pytorch.sh:
```diff
@@ -97,41 +97,37 @@ install_prerequisites
+    #CK/AOTriton - no Flash attention and only corresponding Fused attention backend is enabled
+    #Unfused - Flash and Fused attentions are disabled
+    #Level 1 - run hipBlasLt in auto and unfused modes
+    #Level 3 - run hipBlasLt in all but unfused modes
```
Just "run in..." hipblasLt here was as part of hipblasLt/rocBlas alternative

ipanfilo commented on transformer_engine/common/util/logging.h:
```diff
@@ -113,6 +94,4 @@
-  } while (false)
-
-#endif  // TRANSFORMER_ENGINE_COMMON_UTIL_LOGGING_H_
+  } while (false)
```
cpplint error - no trailing newline

ipanfilo commented on README.rst:
```diff
@@ -107,15 +99,9 @@ Note that some of operator unit tests fail in hipBLASLt config due to limited in
+Pytorch integration pytests under tests/pytorch/ and tests/pytorch/fused_attn/ are supported except for:
 
-Also test_onnx_export.py does not support FP8 dues to absence of custom QDQ operatrs library
+test_onnx_export.py does not support FP8 dues to absence of custom QDQ operators library.
```
Oops. This line is outdated - FP8 is supported with test_onnx_export.py

ipanfilo commented on README.rst:
```diff
@@ -75,15 +75,7 @@ Install TE
-
-If neither of the aforementioned environment variables are set or both of them are set, then both backends are built. If only one of these environment variables is set, then the other backend is not built.
-When both GEMM backends are supported the aforementioned env variables can be used to select which backend to use. If none is set hipBlasLt is used by default. The hipBlasLt backed has not yet supported all the GEMM configurations in the pytorch unit tests. 
+The default installation above supports hipBlasLt in GEMM computation. rocBlas is no longer supported.
```
May be move it to Features on ROCm platform section.

ipanfilo commented on ci/core.sh:
```diff
@@ -31,20 +31,13 @@ if [ $? -eq 0 ]; then
-        test $? -eq 0 || test_run_error "GEMM $_gemm"
-    fi
-done
+_gemm="hipblaslt"
```
Here GEMM tests were separated not to run non-GEMM ones twice: with rocBlas and hipBalsLt. With only one backend they can be run all together or as "nongemm"/"gemm"

wangye805 commented on tests/pytorch/test_cuda_graphs.py:
```diff
@@ -10,7 +10,6 @@
 import pytest
 
 import torch
-from torch.utils.cpp_extension import IS_HIP_EXTENSION
```
Once we remove IS_HIP_EXTENSION, we should expect this file is unchanged compared to NV upstream one. Can you compare it with tests/pytorch/test_cuda_graphs.py in NVTE repo? Thanks

alextmagro replied:
Checked the diff -- There is a difference from NVTE with regards to a "fp8_recipe". Not sure if we can add that to our tests.

wangye805 replied:
Emm, fp8_recipe is another topic and we can hold it off to another PR

wangye805 commented on tests/pytorch/test_fusible_ops.py:
```diff
@@ -10,7 +10,6 @@
 
 import pytest
 import torch
-from torch.utils.cpp_extension import IS_HIP_EXTENSION
```
Same here to make this file consistent with NV upstream

alextmagro replied:
Same as above, there is an fp8_recipe added to several portions of the tests, as well as usage of quantization/quantize variables. This might be the same as some of our fp8 related variables.

ipanfilo replied:
Are you comparing to correct version of upstream?

alextmagro replied:
I was comparing to this upstream commit that Ye shared with me: af7b2b44dd6173c9b3049f306c0773a938feceae

After I rebase, the files should be consistent.

ipanfilo replied:
I believe this is base revision after IFU 2.1. So it should be compared again after the latest dev ported to this PR branch

ipanfilo replied:
So, does the file match upstream version now?

alextmagro replied:
This file and test_cuda_graphs.py now match upstream, except for the AMD copyright.

ipanfilo replied:
Let's remove AMD copyrights there then to reflect the fact that no ROCm specific modifications in those files anymore. I think that is why Ye originally started this comments thread

alextmagro replied:
Done.

wangye805 commented on tests/pytorch/test_sanity.py:
```diff
@@ -923,8 +916,6 @@ def test_sanity_gradient_accumulation_fusion(
     if IS_HIP_EXTENSION:
-        if not use_hipblaslt():
-            pytest.skip("CUDA graph capture not supported with rocBLAS path")
         if dtype in (torch.float16, torch.bfloat16):
```
Our fused attn backends should support cuda graphs now.

wangye805 commented on transformer_engine/common/gemm/rocm_gemm.cu:
```diff
@@ -1402,505 +1068,40 @@ void hipblaslt_gemm(const Tensor *inputA,
 
 void cublas_gemm(const Tensor *inputA, const Tensor *inputB, Tensor *outputD,
                  const Tensor *inputBias, Tensor *outputPreGelu, int m, int n, int k, int lda,
                  int ldb, int ldd, bool transa, bool transb, bool grad,
```
Previously we make transa and transb bool variables to be compatible for both hipblaslt and rocblas backend. After removing rocblas, we can make it the same as upstream NVTE

alextmagro replied:
Should this go alongside the ticket combining cublas_gemm and hipblaslt_gemm?

wangye805 replied:
Yeah, we can hold it to the next PR

Commit:
Fixups and small refactors

ipanfilo commented on README.rst:
```diff
@@ -70,13 +70,12 @@ Install TE
   export NVTE_ROCM_ARCH=gfx942 # CK fused attn only support MI200 and MI300 and fp8 features are only supported on MI300
   
   # Build Platform Selection (optional)
+  # Note: rocBLAS is no longer a supported GEMM backend. hipBLASLt is used for GEMM operations.
```
This paragraph is about build platform, i.e. ROCm vs CUDA. It should not say anything about BLAS library

alextmagro replied:
Removed -- Only reference to rocblas being disabled is around partial gemm support documentation

ipanfilo commented on ci/core.sh:
```diff
@@ -31,13 +31,9 @@ if [ $? -eq 0 ]; then
-_gemm="hipblaslt"
 _exclude="-E Test(.*XTN|.*XNT|.*bf16/.*X.X1|.*fp8.*fp16/.*X1X0|.*fp8.*X.X1|.*fp8/|.*bf8/)"
-check_test_filter $_gemm
-if [ $? -eq 0 ]; then
```
Keep check_test_filter "gemm" and the following if.

ipanfilo commented on ci/pytorch.sh:
```diff
@@ -106,16 +103,16 @@ for _fus_attn in auto flash ck aotriton unfused; do
+    #Level 3 - run in all but unfused modes
     if [ $TEST_LEVEL -ge 3 ]; then
-        test $_gemm = hipblaslt -a $_fus_attn = unfused && continue
+        $_fus_attn = unfused && continue
```
Broken functionality - incorrect syntax

alextmagro replied:
Whoops. Readded "test" to this ifelse block.

ipanfilo commented on README.rst:
```diff
@@ -20,7 +20,7 @@ Feature Support Status
 
 * Activation, cast, fused softmax, layernorm, rmsnorm, transpose, fused rope, fp8 recipe, HipRTC: fully supported
-* GEMM: partially supported with following input/output types: (fp32/fp32), (fp16/fp16), (bf16/bf16), (fp8, bf8/fp16, bf16, fp32)
+* GEMM: partially supported with following input/output types: (fp32/fp32), (fp16/fp16), (bf16/bf16), (fp8, bf8/fp16, bf16, fp32) **Note:** Support for rocBLAS as a GEMM backend has been removed; `hipBLASLt` is the primary backend.
```
The only backend?

alextmagro replied:
Wording fixed

ipanfilo commented on tests/pytorch/test_sanity.py:
```diff
@@ -922,12 +915,6 @@ def test_sanity_gradient_accumulation_fusion(
-    if IS_HIP_EXTENSION:
-        if not use_hipblaslt():
-            pytest.skip("CUDA graph capture not supported with rocBLAS path")
-        if dtype in (torch.float16, torch.bfloat16):
```
Is it supported?

alextmagro replied:
@wangye805 can you confirm this?

ipanfilo replied:
You should answer it based on test results. Also, there is similar filtering in test_numerics.py::test_gpt_cuda_graph which also checks Attention backend, so testing should be done with all - note that default CI config covers auto and unfused only. 
I think removing non-rocBLAS specific code is out of scope of this PR.

alextmagro replied:
Tests passed, so I believe it is supported.

Commit:
Fixed syntax in ci scripts:

- readded missing "test" call
- removed hipblaslt references and replaced with "gemm" where applicable
- readded check_test_filter for "gemm" tests

Fixed changes to README.

ipanfilo commented on ci/pytorch.sh:
```diff
@@ -86,7 +84,7 @@ run_test_config_mgpu(){
 if [ -n "$SINGLE_CONFIG" ]; then
     _gemm=`echo $SINGLE_CONFIG | cut -d- -f1`
     _fus_attn=`echo $SINGLE_CONFIG | cut -d- -f2`
```
SINGLE_CONFIG comes from run_test_job parameter, so it should be just _fus_attn="$SINGLE_CONFIG" now

alextmagro replied:
Changed this, as well as a final _gemm reference, to match with jax.sh.

Commit:
Removed remaining _gemm references from pytorch.sh

Commit:
Readded test_gpt_cuda_graph skip condition -- TODO in #12672

alextmagro commented:
Rebased with IFU v2.1

Commit:
Remove AMD copyrights as file is unchanged from upstream

wangye805 reviewed (approved):
LGTM

wenchenvincent commented:
@wangye805 Don't forget to use "Squash and Merge" next time..

wangye805 commented:
> @wangye805 Don't forget to use "Squash and Merge" next time..

Actually, I did squash and merge. You can see that in our dev branch commit history as this merge only contains one commit

wenchenvincent commented:
> > @wangye805 Don't forget to use "Squash and Merge" next time..
> 
> Actually, I did squash and merge. You can see that in our dev branch commit history as this merge only contains one commit

Oh, sorry, I mixed it up with another PR.
